distributions/aws/infra_configs/scripts/aws/route53.py

```diff
@@ -0,0 +1,159 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Any, Dict
+import boto3
+import logging
+import random
+import string
+
+from botocore.exceptions import ClientError, WaiterError
+from typing import Dict
+
+logger = logging.getLogger(__name__)
+
+
+class Route53HostedZone:
+    """
+    Encapsulates route53 functions.
+    """
+
+    def __init__(
+        self,
+        domain: str,
+        id: str = None,
+        route53_client: Any = None,
+        region: str = "us-east-1",
+    ):
+        """
+        :param route53_client: A Boto3 route53 client.
+        :param domain: domain corresponding to this reference
+        """
+        self.route53_client = route53_client or boto3.client(
+            "route53", region_name=region
+        )
+        self.domain = domain
+        self.id = id
+
+    def create_zone(self) -> dict:
+        randstr = "".join(
+            random.choice(string.ascii_lowercase + string.digits) for _ in range(10)
+        )
+        try:
+            response = self.route53_client.create_hosted_zone(
+                Name=self.domain, CallerReference=f"{self.domain}-{randstr}",
+            )
+        except ClientError:
+            logger.exception(
+                f"failed to create hosted zone for {self.domain} with CallerReference: {randstr}"
+            )
+            raise
+        else:
+            self.id = response["HostedZone"]["Id"].split("/")[-1]
+            logger.info(f"created hosted zone {self.domain} with id {self.id}")
+            return response
+
+    def get_zone(self) -> dict:
+        try:
+            response = self.route53_client.get_hosted_zone(Id=self.id)
+        except ClientError:
+            logger.exception(f"failed to describe hosted zone for {self.domain}")
+            raise
+        else:
+            return response
+
+    def get_name_servers(self) -> list:
+        return self.get_zone()["DelegationSet"]["NameServers"]
+
+    def generate_change_record(
+        self,
+        record_name: str,
+        record_type: str,
+        record_value: list,
+        action: str = "UPSERT",
```

Comment:
what does `UPSERT` action mean here?

Reply:
If a record does not already exist, Amazon Web Services creates it. If a record does exist, Route 53 updates it with the values in the request.

https://docs.aws.amazon.com/Route53/latest/APIReference/API_ChangeResourceRecordSets.html

Reply:
added in comment